ext/mbstring/mbstring.c

```diff
@@ -3560,30 +3560,24 @@ PHP_FUNCTION(mb_send_mail)
 		Z_PARAM_STRING(subject, subject_len)
 		Z_PARAM_STRING(message, message_len)
 		Z_PARAM_OPTIONAL
-		Z_PARAM_ZVAL(headers)
-		Z_PARAM_STR(extra_cmd)
+		Z_PARAM_STR_OR_ARRAY_HT_OR_NULL(str_headers, headers_ht)
+		Z_PARAM_STR_OR_NULL(extra_cmd)
 	ZEND_PARSE_PARAMETERS_END();
 
 	/* ASCIIZ check */
 	MAIL_ASCIIZ_CHECK_MBSTRING(to, to_len);
 	MAIL_ASCIIZ_CHECK_MBSTRING(subject, subject_len);
 	MAIL_ASCIIZ_CHECK_MBSTRING(message, message_len);
-	if (headers) {
-		switch(Z_TYPE_P(headers)) {
-			case IS_STRING:
-				tmp_headers = zend_string_init(Z_STRVAL_P(headers), Z_STRLEN_P(headers), 0);
-				MAIL_ASCIIZ_CHECK_MBSTRING(ZSTR_VAL(tmp_headers), ZSTR_LEN(tmp_headers));
-				str_headers = php_trim(tmp_headers, NULL, 0, 2);
-				zend_string_release_ex(tmp_headers, 0);
-				break;
-			case IS_ARRAY:
-				str_headers = php_mail_build_headers(Z_ARRVAL_P(headers));
-				break;
-			default:
-				zend_argument_value_error(4, "must be of type string|array|null, %s given", zend_zval_type_name(headers));
-				RETURN_THROWS();
-		}
+
+	if (str_headers) {
+		tmp_headers = zend_string_init(ZSTR_VAL(str_headers), ZSTR_LEN(str_headers), 0);
+		MAIL_ASCIIZ_CHECK_MBSTRING(ZSTR_VAL(tmp_headers), ZSTR_LEN(tmp_headers));
```

Comment:
Drive-by note: I think the MAIL_ASCIIZ_CHECK_MBSTRING calls above are corrupting the argument strings :(